webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/batch/GitGC.scala

```diff
@@ -0,0 +1,116 @@
+* the following Additional permissions:
+* Notwithstanding to the terms of section 5 (5. Conveying Modified Source
+* Versions) and 6 (6. Conveying Non-Source Forms.) of the GNU General
+* Public License version 3, when you create a Related Module, this
+* Related Module is not considered as a part of the work and may be
+* distributed under the license agreement of your choice.
+* A "Related Module" means a set of sources files including their
+* documentation that, without modification of the Source Code, enables
+* supplementary functions or services in addition to those offered by
+* the Software.
+*
+* Rudder is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+*
+*************************************************************************************
+*/
+
+package com.normation.rudder.batch
+
+import com.normation.rudder.domain.logger.GitRepositoryLogger
+import com.normation.rudder.git.GitRepositoryProvider
+import com.normation.utils.CronParser._
+
+import cron4s.CronExpr
+import org.eclipse.jgit.lib.ProgressMonitor
+import org.joda.time.Duration
+
+import zio._
+import zio.clock.Clock
+import com.normation.errors.IOResult
+import com.normation.zio._
+
+
+/**
+ * A scheduler which run a git gc every day.
+ * It use a cron-like config parsed by cron4s.
+ *
+ * Code derived from https://github.com/centic9/jgit-cookbook/blob/master/src/main/java/org/dstadler/jgit/porcelain/CollectGarbage.java
+ */
+class GitGC(
+    gitRepo: GitRepositoryProvider
+  , optCron: Option[CronExpr]
+) {
+
+  val logger = GitRepositoryLogger
+
+  // this class is necessary to report information about gitgc processing that can be long
+  class LogProgressMonitor extends ProgressMonitor {
+    def start(totalTasks: Int): Unit = {
+      logger.debug(s"Start gitgc on ${gitRepo.rootDirectory.name} for ${totalTasks} tasks")
+    }
+
+    def beginTask(title: String, totalWork: Int): Unit = {
+      logger.debug(s"Start gitgc on ${gitRepo.rootDirectory.name} for ${title}: ${totalWork}")
+    }
+
+    def update(completed: Int): Unit = {
+      logger.trace(s"gitgc completed on ${gitRepo.rootDirectory.name}: ${completed}")
+    }
+
+    def endTask(): Unit = {
+      logger.trace(s"gitgc on ${gitRepo.rootDirectory.name}: task done")
+    }
+
+    def isCancelled = false
+  }
+
+  // must not fail, will be in a cron
+  val gitgc: UIO[Unit] = for {
+    t0 <- currentTimeMillis
+    _  <- gitRepo.semaphore.withPermit(IOResult.effect {
+           gitRepo.git.gc().setProgressMonitor(new LogProgressMonitor()).call
+         }).catchAll(err => logger.error(s"Error when performing GitGC on ${gitRepo.rootDirectory.name}: ${err.fullMsg}"))
```

Comment:
we should decide between "gitgc", "GitGC" in logs. I would prefer `git-gc`, but anything goes.